### DISCUSSION TITLE
Go SDK: support embedding structs

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12327] [Ideas]
It would be cool if the Go SDK supported embedded another type. For example:

```golang
type *SuperDirectory {
  *dagger.Directory
}

// Extend the core directory type with an extra function
func (dir *SuperDirectory) Hello() string {
 return "world"
}
```

Since `SuperDirectory` embeds `dagger.Directory`, it inherits all the methods of the embedded type. But this embedding does not carry into the corresponding Dagger module. So, even though I can call `SuperDirectory.Entries()` in Go via embedding, I cannot call `SuperDirectory.Entries()` via the Dagger API.

**Comment by sagikazarmark** [comment]
One challenge I see: `SuperDirectory.WithFile()` would still return a `*Directory` (unless the code generator magically take care of that)

**Comment by jedevc** [comment]
I think this would potentially be a breaking change (since new methods might clash). That said, I think I agree, I like the idea of doing this.

> One challenge I see: `SuperDirectory.WithFile()` would still return a `*Directory` (unless the code generator magically take care of that)

I think we should have the code generator magically take care of this.

Potentially, if we do this "inheritance"/"composition", we should implement this at the core API level, instead of delegating to each SDK. Maybe something like `.Extends()` or similar - overridden methods get delegated to the module.

If we do this, I think it would make sense to allow something similar in python/typescript/etc - it's pretty useful functionality IMO, so it should be available in all SDKs.

**Comment by sagikazarmark** [comment]
Something that I often do is expose certain parts of the `Container` type. Mostly `WithEnvVariable`. Exposing partial APIs with embedding may not work, but I though I'd mention that use case.

**Comment by b4nst** [comment]
> Maybe something like .Extends() or similar - overridden methods get delegated to the module.

One - potential - benefit or caveat of that over delegating to SDK is we could keep the magic dagger is handling for us right now.
E.G. When doing something like

```go
type SuperDirectory struct {
  //+private
  *dagger.Directory
}
```

Currently it seems dagger is handling context and errors under the hood.

```go
func (s *SuperDirectory) Digest() string {
  return s.Directory.Digest()
}
```

I'm still unsure if I find it confusing or super convenient

**Comment by nipuna-perera** [comment]
Needed this today. We have a module that builds container images that adhere to our standards. We are having to manually expose a whole bunch of the `Container` functions to accomodate that. Embedding it would have been a much better experience.
